scst/src/scst_targ.c

```diff
@@ -3664,11 +3664,9 @@ static int scst_pre_xmit_response1(struct scst_cmd *cmd)
 		/*
 		 * Those counters protect from not getting too long processing
 		 * latency, so we should decrement them after cmd completed.
+		 *
+		 * @cmd processing for SCST device is complete.
 		 */
-		smp_mb__before_atomic_dec();
-		WARN_ON_ONCE(!cmd->owns_refcnt);
-		cmd->owns_refcnt = false;
```

Comment:
`smp_mb__before_atomic_dec();` applies to `atomic_dec(&cmd->tgt_dev->tgt_dev_cmd_count);` and hence should be moved to just in front of where the `atomic_dec()` has been moved.

Reply:
Before this patch `smp_mb__before_atomic_dec();` used to make sure that `cmd->owns_refcnt = false;` is perceived to be set before the reference counter is decremented. But once the decrementing has been moved, I'm not sure we need to use that barrier anymore.

1. Moving the barrier in front of where the decrementing has been moved doesn't make any sense, because there is nothing already before `atomic_dec()` that we want to 'guard'. Taking into account the fact that the check for `cmd->owns_refcnt` occurs before decrementing: `WARN_ON_ONCE(cmd->owns_refcnt);`.
2. The `cmd->owns_refcnt` and `cmd->tgt_dev->tgt_dev_cmd_count` have independent semantic so it doesn't matter if `cmd->owns_refcnt` will be set after `cmd->tgt_dev->tgt_dev_cmd_count` was decremented.

So why we just cannot remove the usage of this barrier?

Reply:
It is not clear to me where the worries about the `smp_mb__before_atomic_dec()` come from since that macro is translated into `do { } while (0)` for x86 CPUs?

The barrier is still necessary to prevent that any store operations that precede the atomic_dec() are observed by another CPU after the atomic_dec() has finished. This can't happen on an x86 CPU but can happen on CPUs that allow store operations to be observed by another CPU in another order than program order (ARM, PowerPC).

Reply:
Okay, thanks for the explanation.

I have moved the barrier in front of new atomic_dec() location.